### DISCUSSION TITLE
boundaries of subdomains

### DISCUSSION THREAD

**Discussion by gatling-nrl** [kinnala/scikit-fem#798] [Ideas]
I propose extending the `class Mesh` to provide some ability to return the boundary of subdomains, even for those cases where the subdomain boundary is not used as a boundary condition.

https://github.com/kinnala/scikit-fem/blob/71eb03e9baee4deab12c9fd91a04118bde3838ec/docs/examples/ex13.py#L62-L63

In Ex13 you use the boundary of the domain. But suppose the domain had an internal boundary between two subdomains. Would the following changes be a good idea?

```python
Mesh.boundary_cells(self, subdomain=None, external=False)  # doesn't exist at present
Mesh.boundary_facets(self, subdomain=None)
Mesh.boundary_edges(self, subdomain=None)  # aren't facets and edges the same thing?
Mesh.boundary_nodes(self, subdomain=None)
```

Where `external=False` specifies you want the cells with edges on the boundary and non-boundary nodes inside the subdomain and `external=True` means cells with edges on the boundary and non-boundary nodes outside the subdomain.

Would these additions make it trivial to extend ex13 to measurements of flux across internal subdomain boundaries?

Is it worth extending the `Mesh.*_satisfying()` functions to have subdomain awareness?

If these are good ideas, I'll start a branch for them.

**Comment by gdmcbain** [comment]
This isn't exactly the same as 'internal boundaries in `with_boundaries`' #683 but some of that might help.

**Comment by kinnala** [comment]
> ```python
> Mesh.boundary_edges(self, subdomain=None)  # aren't facets and edges the same thing?
> ```

> Nomenclature: In scikit-fem, edges exist only for three-dimensional meshes so that facets are something always shared between two elements of the mesh. In particular, we refer to the edges of triangular and quadrilateral meshes as facets.

**Comment by kinnala** [comment]
> ```python
> Mesh.boundary_facets(self, subdomain=None)
> ```
This idea is fine. However, implementing this may not be completely straightforward.

The reason why we are able to find those boundary faces in a reasonable time is that the second row of `f2t` contains -1 if the facet is on the boundary. This is done in `_build_entities` and done only once for the whole mesh.

I think the easiest way to hack this together would be to create another `Mesh` object consisting of only the subdomain elements, using `Mesh._reix` (or similar), and then call `Mesh.boundary_facets()` on the resulting object.

In particular, implementing `boundary_edges(subdomain='blah')` for `MeshTet1` and `MeshHex1` will be otherwise pretty painful. However, I think `boundary_edges(subdomain='blah')` may not be needed if the use case is to calculate the flux?

**Reply by kinnala** [reply]
Here is one idea for the implementation:
```python
m = MeshTri().refined(4).with_subdomains({'mid': lambda x: x[0] <= 0.5})
v, c = np.unique(m.t2f[:, m.subdomains['mid']].flatten(), return_counts=True)
facets = v[np.argwhere(c == 1)].flatten()
```
I tried drawing `facets`:
```python
In [29]: basis = Basis(m, ElementTriP1())                                     
In [30]: dofs = basis.get_dofs(facets)                                        
In [31]: u = basis.zeros()                                                    
In [32]: u[dofs] = 1                                                          
In [33]: m.draw(point_data={'u': u}).show() 
```
![Kuvakaappaus - 2021-11-18 18-22-43](https://user-images.githubusercontent.com/973268/142455090-9ef02ef4-6743-4841-846a-bb36964fb8cc.png)

**Reply by kinnala** [reply]
Maybe use nonzero instead of argwhere.

**Comment by kinnala** [comment]
> ```python
> Mesh.boundary_cells(self, subdomain=None, external=False)  # doesn't exist at present
> ```
Will this include only elements having a facet on the boundary or also elements that have only one node/edge at the boundary?

The former can be done by a simple `m.f2t[0, m.boundary_facets()]`. The latter is somewhat nontrivial to do efficiently.

Edit: We should call it `boundary_elements` so that it is in line with `elements_satisfying` etc.

**Reply by gatling-nrl** [reply]
That's a hard question to answer. Perhaps a parameter to control that? I believe that ex13 would want only elements with a facet on the boundary, right?

How does `f2t` select only the elements of one of the subdomains?

**Reply by kinnala** [reply]
What would be the purpose of `boundary_cells/elements`? I cannot imagine any use case.

**Reply by gatling-nrl** [reply]
Suppose I have Laplace equation with coefficient like this. Yellow is orders of magnitude different from blue. Like an metal electrode in some kind of non-homogenous semi-conductive fluid. The "current" to the electrode in this case is the integral of  (du*nhat)/sigma  around the boundary between domains. If sigma is piecewise constant (ElementTriP0 for example) don't i need to get the elements in the blue domain that are part of the boundary with the yellow?

![image](https://user-images.githubusercontent.com/38136423/142474544-ebeda08a-e088-414e-9479-4eef42f5e14a.png)

**Reply by kinnala** [reply]
Possibly, but you can also use `InteriorFacetBasis` to integrate over the interface. I can give you an example shortly.

**Reply by kinnala** [reply]
Something like this:
```python
from skfem import *
from skfem.models import laplace, unit_load
import numpy as np
m = MeshTri().scaled(4.0).translated((-2.0, -2.0)).refined(4).with_subdomains({'circle': lambda x: (x[0] ** 2 + x[1] ** 2) <= 1.})
v, c = np.unique(m.t2f[:, m.subdomains['circle']].flatten(), return_counts=True)
facets = v[np.nonzero(c == 1)[0]]
basis = Basis(m, ElementTriP1())
A = laplace.assemble(basis)
f = unit_load.assemble(basis)
u = solve(*condense(A, f, I=m.interior_nodes()))
ifbasis = InteriorFacetBasis(m, ElementTriP1(), facets=facets)
@Functional
def port_flux(w):
    from skfem.helpers import dot, grad
    return dot(w.n, grad(w['u']))
current = asm(port_flux, ifbasis, u=u)
```
You can pass `side=0` or `side=1` to `InteriorFacetBasis`. In theory, `side` decides the side of every facet where you choose the element for which the derivative is taken. However, if you want `side=0` to always correspond to, e.g., outside the interface, then `m.t` must be ordered in a specific way. Unfortunately, there is no built-in routine for that.

**Reply by kinnala** [reply]
Alright, after thinking about the above, it probably doesn't make sense if the direction of the normal changes from one facet to another. So some additional reorienting is required. I'll give it a thought and come back with another snippet.

**Reply by gatling-nrl** [reply]
```
@Functional
def port_flux(w):
    from skfem.helpers import dot, grad
    return dot(w.n, grad(w['u']))
```
What does `w.n` contain?

**Reply by gdmcbain** [reply]
The unit outward normal.

**Reply by gdmcbain** [reply]
The snippet implements **n** &sdot; &nabla; *u*.

**Reply by gatling-nrl** [reply]
I thought that's what it was doing, but I wasn't sure normal to ... what? A boundary facet? Defined at the midpoint of the facet?

**Reply by gatling-nrl** [reply]
Found this, but its still not that clear to me what's going on with it.

https://github.com/kinnala/scikit-fem/blob/71eb03e9baee4deab12c9fd91a04118bde3838ec/skfem/element/element_global.py#L159-L166

**Reply by gdmcbain** [reply]
> I thought that's what it was doing, but I wasn't sure normal to ... what? A boundary facet? 

Yes, this form is only evaluated on facet bases.

> Defined at the midpoint of the facet?

The elements here are affine rather than isoparametric so the facets are rectilinear and the normal constant over the facet.

**Reply by kinnala** [reply]
> Found this, but its still not that clear to me what's going on with it.
> 
> https://github.com/kinnala/scikit-fem/blob/71eb03e9baee4deab12c9fd91a04118bde3838ec/skfem/element/element_global.py#L159-L166

This is a different normal. The normal used inside the forms is created in `MappingAffine.normals`.

However, my snippet has an issue because it doesn't verify that the normal is always pointing outwards w.r.t the subdomain. For the normal to be outwards I need to somehow orient the facets belonging to the interface.

**Reply by kinnala** [reply]
So I did some kind of manual orientation of `w.n` inside the form and now it gives a reasonable result:
```python
from skfem import *
from skfem.models import laplace, unit_load
import numpy as np

m = MeshTri().scaled(4.0).translated((-2.0, -2.0)).refined(5).with_subdomains({'circle': lambda x: (x[0] ** 2 + x[1] ** 2) <= 1.})
v, c = np.unique(m.t2f[:, m.subdomains['circle']].flatten(), return_counts=True)
facets = v[np.nonzero(c == 1)[0]]
basis = Basis(m, ElementTriP1())
A = laplace.assemble(basis)

@LinearForm
def load(v, w):
    return (np.abs(w.x[0]) <= .5) * (np.abs(w.x[1]) <= .5) * v

f = load.assemble(basis)
u = solve(*condense(A, f, I=m.interior_nodes()))
ifbasis = InteriorFacetBasis(m, ElementTriP1(), facets=facets)

@Functional
def port_flux(w):
    from skfem.helpers import dot, grad
    # orient the normal vector
    n = np.sign(dot(w.n, w.x)) * w.n
    return dot(n, grad(w['u']))

current = asm(port_flux, ifbasis, u=u)
# current = -1.0326412218753906
```
However, since we know that this `ifbasis` comes from a closed loop there should be a better way of orienting the facets.

However, back to the original problem, isn't there a better way to calculate the same quantity using the Divergence Theorem and some domain integral?

**Reply by gatling-nrl** [reply]
> However, back to the original problem, isn't there a better way to calculate the same quantity using the Divergence Theorem and some domain integral?

There doesn't seem to be. I'm dealing with an "electrode" where the conductivity of the electrode might be `>1e6` and of the surrounding medium `1`. This makes the gradients inside the electrode very small (zero for an ideal conductor). It seems computationally safer to me to work with the gradients in the area immediately surrounding the electrode. It isn't even interesting to my problem to model the inside of the electrode. It is sufficient to simply place a boundary condition at the electrode edge. Of course, this is one of the things I haven't gotten right yet because ideally this boundary condition is "isopotential with constant value of integrated flux" (the current source).

I'm trying to get a handle on how `skfem.FacetBasis` works today.

**Reply by gdmcbain** [reply]
> Of course, this is one of the things I haven't gotten right yet because ideally this boundary condition is "isopotential with constant value of integrated flux" (the current source).

Here's one way to think about this problem which might be instructive. (And even practicable.) Consider a different problem in which a uniform potential is prescribed on the electrode. Then from the solution of this problem integrate the flux. We now have a function which maps potential to current.  If we can solve this scalar function for the potential that gives the desired current, we have solved the original problem.

**Reply by gatling-nrl** [reply]
My fall back plan is essentially that exactly. Since generalized ohms law is governing equation, for any fixed conductivity the current will be a linear function of voltage. So the worst case is to solve the problem 2 times to get that linear function, and then a 3rd time to get the specified current. I have the luxury of relatively simple domain shapes and meshes, but I'm running many thousands of variations so adding any kind of non-linear search to the solution is probably to expensive. But I've had this 3-solution penalty in the back of my head for a while now.

I can also work around those huge ratios (`1:1e6`) in conductivity by putting boundaries around the `1e6` regions and ignoring them. However I still see problems with ratios of `1:100` which I don't really consider to be "huge". Which makes me think that the contrast in conductivity is not the reason the solution is failing. And sorry, I seem to be getting threads crisscrossed now. I don't remember where we were talking about contrast problems.

**Reply by gdmcbain** [reply]
> There doesn't seem to be. I'm dealing with an "electrode" where the conductivity of the electrode might be >1e6 and of the surrounding medium 1. This makes the gradients inside the electrode very small (zero for an ideal conductor). 

The domain integral then would be over the "surrounding medium". This is much more accurate than a boundary integral. See ex13 and the ANZIAM paper.

**Reply by gdmcbain** [reply]
> My fall back plan is essentially that exactly. Since generalized ohms law is governing equation, for any fixed conductivity the current will be a linear function of voltage. So the worst case is to solve the problem 2 times to get that linear function, and then a 3rd time to get the specified current. I have the luxury of relatively simple domain shapes and meshes, but I'm running many thousands of variations so adding any kind of non-linear search to the solution is probably to expensive.

Not necessarily.  This kind of technique (&lsquo;Schur complement&rsquo;) can be used on a large scale; e.g. one approach to incompressibility in fluid mechanics is to construct a function that for a given pressure field computes the divergence field, then one &lsquo;solves&rsquo; that function to find the pressure that gives zero divergence everywhere.  That function is affine (linear plus constant) so can be converted to linear and solved with conjugate gradient, MINRES, &c..  This is what's done in ex30.

https://github.com/kinnala/scikit-fem/blob/be9642f06927ba402fe916abba061d110759b095/docs/examples/ex30.py#L96-L109

**Reply by kinnala** [reply]
> I'm trying to get a handle on how `skfem.FacetBasis` works today.

Facet integration is implemented differently than in almost any FEM codes I have read. I learned the technique from a colleague of mine several years ago (credits in the README). Consider a triangle mesh. The idea is to first take the reference element of the boundary cell (line) and create quadrature points. Then we map those points to the facets of the triangle mesh using MappingAffine.G and finally back to the reference triangle using MappingAffine.invF. Now we have a set of quadrature points on the correct facet of the reference triangle and can evaluate the local basis.

The benefit is that we automatically get gradient information which is normally a challenge (based on many questions in scicomp StackExchange). I guess the downside is that by mapping back and forth you do a bit of extra work.

However, you can think of it as follows: by changing Basis to FacetBasis you are changing only the integration domain to the boundary. Most of the forms are compatible with both with the exception that FacetBasis includes w.n for outward normal.

**Reply by gatling-nrl** [reply]
![image](https://user-images.githubusercontent.com/38136423/144082081-688e099a-055b-4cbe-8d55-64396bf505bb.png)

Like that?

**Reply by kinnala** [reply]
Yes.
